### PR TITLE
cli/sql: further polish statement statistics

### PR DESCRIPTION
**Pull request by knz**
The following rules are applied:

- the breakdown is only displayed if the total execution time exceeds
  1ms. We assume that (human) users don't care about the breakdown if a
  statement is faster than that.

- if `verbose_timings` is unset (default), the server-side latencies
  are all grouped together (the "service latency") under a single
  `execution` label.

- the user can invoke `\set verbose_timings` to get the full
  breakdown.

Example outputs follow.

The breakdown is omitted because the total time is under 1ms:

```
root@127.0.0.1:44820/defaultdb> select 1;
  ?column?
------------
         1
(1 row)

Time: 0.000s total
```

A query is simple to parse/plan, but takes a while to execute because
it produces a lot of data:

```
root@127.0.0.1:44820/defaultdb> select generate_series(1,10000);
...
(10000 rows)

Time: 0.013s total (execution 0.004s / network 0.009s)
```

A more complex statement:

```
root@127.0.0.1:44820/defaultdb> create table t(x int);
CREATE TABLE

Time: 0.003s total (execution 0.002s / network 0.001s)
```

Release note (cli change): The display of statement timings in the SQL
shell (`cockroach sql` and `cockroach demo`) has been further simplified.

### CODE REVIEW

**Comment by jseldess** [comment]
Thank you, @knz! I like rule 1, but I'm not sure about rule 2. I think only showing the parts larger than 1% of the total might be confusing. I'd rather show the parts that are always involved (plan, exec, network) and find a way to indicate that they are negligible if less than 1%. perhaps even just `<1%`? Thoughts?

**Review by arulajmani — COMMENTED** [review]
That was quick! I'm slightly thrown off by these things not adding up to 100%. IMO swallowing <1% adds more confusion than having an "other" bucket. Would it be too verbose if we printed all the buckets regardless of how small they were?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54751#-:-MI-dfe41AQKtluo0vIB:b-xrnhvt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @jseldess, and @RaduBerinde)

**Comment by knz** [comment]
> Would it be too verbose if we printed all the buckets regardless of how small they were?

```
root@127.0.0.1:44820/defaultdb> select generate_series(1,10000);
...
(10000 rows)

Time: 0.012s total (parse 0.1% / plan 0.6% / exec 29.4% / background 0.5% / network 69.4%)
```

I find it a little bit verbose, yes.

**Comment by arulajmani** [comment]
What if we only disambiguated between server and network?

**Comment by irfansharif** [comment]
To add to the bikeshed, I'd be happy with "other" or "background" or any arbitrary grouping if I could toggle between the latencies grouped together vs. having the full breakdown made available to me.

**Comment by irfansharif** [comment]
To that end, I don't think as an end user I really care all that much about "parse", "plan", etc. "exec" and "network" yes, everything else is probably CRDB eng speak. I'd be surprised if users expected "parse" and "plan" to be separated out from "exec".

**Comment by knz** [comment]
> I could toggle between the latencies grouped together vs. having the full breakdown made available to me.

that remains available with `\set verbose_timings on` in any case

**Comment by knz** [comment]
ok Irfan's point and Arul's point together:

- when `verbose_timings` is unset (default), we only break down between `server` and `network`
- if `verbose_timings` is set we show the details

would that work?

**Comment by irfansharif** [comment]
That works, though I don't like "server". I think exec/execution is better+more instructive. It's generic enough to group all these "other" categories into it when it's non-verbose.

**Comment by knz** [comment]
ok done, PTAL

**Comment by arulajmani** [comment]
Im on board with the new approach :+1:

**Comment by knz** [comment]
Changed to display the latencies as absolute time instead of percentages, as recommended by @bdarnell 

PTAL

File: pkg/cli/sql_util.go [hunk 1]
```diff
@@ -989,23 +999,36 @@ func maybeShowTimes(
...
+		//
+		// Note: we omit the report of percentages for queries that
+		// last for a millisecond or less. This is because for such
+		// small queries, the detail is just noise to the human observer.
+		sep := " ("
+		maybeReport := func(label string, lat time.Duration) {
+			fmt.Fprintf(w, "%s%s %.3fs", sep, label, lat.Seconds())
+			sep = " / "
+		}
+		maybeReport("execution", serviceLat)
```
Comment (irfansharif) [inline pkg/cli/sql_util.go:1026]: What's maybe about it?

**Review by irfansharif — APPROVED** [review]

File: pkg/cli/sql_util.go (hunk 1 above)
Comment (knz) [reply pkg/cli/sql_util.go]: Fixed. Thanks

**Review by arulajmani — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54751)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @jseldess, @knz, and @RaduBerinde)

___
*[pkg/cli/sql_util.go, line 1028 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/54751#-MI0-P0t-E3KIqNWrPaB:-MI0-P0t-E3KIqNWrPaC:b-28cei4) ([raw file](https://github.com/cockroachdb/cockroach/blob/735f28615f22805831a26b34cdc303cd42522db7/pkg/cli/sql_util.go#L1028)):*
> ```Go
> 		reportTiming("execution", serviceLat)
> 		reportTiming("network", networkLat)
> 		if sep != "( " {
> ```

nit: This can be cleaned up now too.

**Comment by knz** [comment]
TFYRs!

bors r=irfansharif,arulajmani,jordanlewis,awoods187,bdarnell,jseldess,RaduBerinde,Annebirzin,ricardocrdb

**Comment by RaduBerinde** [comment]
Looks great to me. Could we use `ms` instead of `s` for sub-second times?

**Comment by knz** [comment]
> Could we use ms instead of s for sub-second times?

I found experimentally that the mix of units makes the output confusing and excessively hard to read.

We'd need all 3 units to be the same. But then if one is large and one is small they would not work well. So the conditional would be rather complex to implement.

If you have a concrete code suggestion I'll take it

**Comment by knz** [comment]
Also remember this output is for the human user. Humans don't care about sub-millisecond differences.

You can get all the details you want with `\set verbose_timings` too

**Comment by RaduBerinde** [comment]
We can condition the `ms` vs `s` on the total. If total is less than 1s, use ms everywhere. We would not show decimal points for `ms`.

**Comment by knz** [comment]
sure why not

bors r-

File: pkg/cli/sql_util.go [hunk 2]
```diff
@@ -989,23 +999,33 @@ func maybeShowTimes(
...
-		fmt.Fprintf(w, " total (exec %s / net %s / other %s)\n",
-			execLatency, networkLatency, serviceLatency-execLatency)
-	} else {
-		// Simplified display: just show percentages.
-		totalSeconds := clientSideQueryTime.Seconds()
-		fmt.Fprintf(w, " total (exec %.1f%% / net %.1f%% / other %.1f%%)\n",
-			execLatency.Seconds()*100/totalSeconds,
-			networkLatency.Seconds()*100/totalSeconds,
-			(serviceLatency-execLatency).Seconds()*100/totalSeconds)
+		fmt.Fprintf(w, " (parse %s / plan %s / exec %s / other %s / network %s)\n",
```
Comment (rafiss) [inline pkg/cli/sql_util.go:1013]: would you consider it in-scope for this PR to always include the verbose timings if in EXPLAIN ANALYZE?

Comment (knz) [reply pkg/cli/sql_util.go:1013]: huh not really no

Comment (rafiss) [reply pkg/cli/sql_util.go:1013]: fine by me. it's something to consider for later based on the feedback from customer-support. (or maybe we should just make sure the CS team knows about `\set verbose_timings`)

Comment (knz) [reply pkg/cli/sql_util.go:1013]: the issue is that this would require client-side parsing which is not any more part of the "presentation" logic.

**Comment by knz** [comment]
added a commit to use milliseconds for sub-second queries, as follows:

```
root@127.0.0.1:30822/defaultdb> select pg_sleep(0.1);
  pg_sleep
------------
    true
(1 row)

Time: 104ms total (execution 104ms / network 0ms)

root@127.0.0.1:30822/defaultdb> select pg_sleep(1);
  pg_sleep
------------
    true
(1 row)

Time: 1.003s total (execution 1.003s / network 0.000s)
```

@RaduBerinde does that work?

**Comment by RaduBerinde** [comment]
Very nice, thanks! Code change LGTM too.

**Comment by knz** [comment]
bors r=irfansharif,arulajmani,jordanlewis,awoods187,bdarnell,jseldess,RaduBerinde,Annebirzin,ricardocrdb

**Pull request merged**